### PR TITLE
Add nitf

### PR DESCRIPTION
**Pull request by riverloopsec**
This PR adds the NITF (National Image Transition Format) format specification.

Please see attached ZIP that has a selected set of NITF images collected from around the internet and carefully selected to have good coverage for testing purposes -
[corpora.zip](https://github.com/kaitai-io/kaitai_struct_formats/files/4743796/corpora.zip). In addition, https://github.com/codice/imaging-nitf/tree/master/shared-test-resources/src/main/resources is another repository of these files.

We have conducted extensive analysis of the security of this format using this as one of our methods for exploring it.

Credit to @TACIXAT for implementation.

### CODE REVIEW

File: image/nitf.ksy [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+    - ntf.r0
+    - ntf.r1
+    - ntf.r2
+    - ntf.r3
+    - ntf.r4
+    - ntf.r5
+  ks-version: 0.9
+  encoding: UTF-8
+  endian: be
+  license: CC-BY-SA-4.0
```
[the remark this answers is not in the document]
Comment (rmspeers) [reply image/nitf.ksy:16]: Updated to MIT.

**Comment by rmspeers** [comment]
@KOLANICH can you look at this we ran into when updating? This is produced by KSV, no errors on KSC.

```
Compilation OK
... processing image/nitf.ksy 0
...... loading nitf.rb
Classes loaded OK, main class = Nitf
Traceback (most recent call last):
	8: from /usr/local/bin/ksv:23:in `<main>'
	7: from /usr/local/bin/ksv:23:in `load'
	6: from /Library/Ruby/Gems/2.6.0/gems/kaitai-struct-visualizer-0.7/bin/ksv:53:in `<top (required)>'
	5: from /Library/Ruby/Gems/2.6.0/gems/kaitai-struct-visualizer-0.7/lib/kaitai/struct/visualizer/visualizer.rb:13:in `run'
	4: from /Library/Ruby/Gems/2.6.0/gems/kaitai-struct-visualizer-0.7/lib/kaitai/struct/visualizer/parser.rb:32:in `load'
	3: from /var/folders/11/nn476p9172xb4__85r_ncw6h0000gn/T/d20200703-41609-v4w6iu/nitf.rb:28:in `_read'
	2: from /var/folders/11/nn476p9172xb4__85r_ncw6h0000gn/T/d20200703-41609-v4w6iu/nitf.rb:28:in `times'
	1: from /var/folders/11/nn476p9172xb4__85r_ncw6h0000gn/T/d20200703-41609-v4w6iu/nitf.rb:31:in `block in _read'
/var/folders/11/nn476p9172xb4__85r_ncw6h0000gn/T/d20200703-41609-v4w6iu/nitf.rb:229:in `_read': undefined method `total_size' for nil:NilClass (NoMethodError)
```

File: image/nitf.ksy [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+        type: str
+        size: 5
+      - id: fscpys
+        type: str
+        size: 5
+      - id: encryp
+        type: encrypt
+      - id: fbkgc
+        size: 3
+      - id: oname
```
[the remark this answers is not in the document]
Comment (rmspeers) [reply image/nitf.ksy]: Updated these for many of the items. Full descriptions in the source doc linked are long and prose, so included some shortened ones where I felt the field could use more context.

File: image/nitf.ksy [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+      - id: desdata
+        type: str
+        size: >
+          total_size - (2 + 25 + 2 + des_base.declasnfo.total_size + 
+            (tre_ofl ? 6 + 3 : 0) + 4 + desshl.to_i)
+  data_sub_header_tre:
+    seq:
+      - id: des_base
+        type: data_sub_header_base
+      - id: desoflw
```
Comment (KOLANICH) [inline image/nitf.ksy:681]: 1. should these 2 fields be separated in a struct?
2. please don't merge words in variable names.

File: image/nitf.ksy [hunk 4]
```diff
@@ -0,0 +1,759 @@
...
+        -orig-id: udid
+        type: u1
+        if: user_def_img_data_len.to_i > 2
+        repeat: expr
+        repeat-expr: user_def_img_data_len.to_i - 3
+      - id: image_extended_sub_header
+        type: tre_header
+  band_info:
+    seq:
+      - id: n_band_representation
```
Comment (KOLANICH) [inline image/nitf.ksy:432]: should `n_band_` be ommitted?

File: image/nitf.ksy [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+        -orig-id: im
+        contents: 'IM'
+      - id: image_id_1
+        -orig-id: iid1
+        type: str
+        size: 10
+      - id: image_date_time
+        -orig-id: idatim
+        type: date_time
+        doc: 'UTC time of image acquisition in the format CCYYMMDDhhmmss: CC century, YY last two digits of the year, MM month, DD day, hh hour, mm minute, ss second'
```
Comment (KOLANICH) [inline image/nitf.ksy:279]: The format of the time string docs should be moved into the struct representing time.

File: image/nitf.ksy (hunk 3 above)
Comment (rmspeers) [reply image/nitf.ksy:681]: `desoflw` is one from the original data, I'll go move these to orig-id and put instead other ids.

File: image/nitf.ksy (hunk 5 above)
Comment (rmspeers) [reply image/nitf.ksy:279]: Moved

File: image/nitf.ksy (hunk 4 above)
Comment (rmspeers) [reply image/nitf.ksy:432]: I had it in there per the doc naming (pg 97 of reference), but have removed.

**Comment by rmspeers** [comment]
FYI for a basic single image test see attached rgb16.ntf:

[rgb16.ntf.zip](https://github.com/kaitai-io/kaitai_struct_formats/files/4870917/rgb16.ntf.zip)

File: image/nitf.ksy [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+      - type: str
+        size: 14
+  encrypt:
+    seq:
+      - type: str
+        size: 1
+  clasnfo:
+    instances:
+      total_size: 
+        value: (1 + 2 + 11 + 2 + 20 + 2 + 8 + 4 + 1 + 8 + 43 + 1 + 40 + 1 + 8 + 15)
```
[the remark this answers is not in the document]
Comment (generalmimon) [reply image/nitf.ksy:163]: @riverloopsec @rmspeers As @KOLANICH points out, there's really no need for such a lengthy and error-prone summation. Exactly this problem is addressed by the [`sizeof` operator](https://github.com/kaitai-io/kaitai_struct/issues/84) (make sure you're using the latest **unstable** compiler version available from https://kaitai.io/#download, or you can use https://ide.kaitai.io/devel/, which has constantly the latest development version out of the box).

I saw that you're using it somewhere like `reclasnfo.total_size`. You can replace this usage with either
 * [type-based sizeof](https://github.com/kaitai-io/kaitai_struct/issues/84#issuecomment-486645268): **`sizeof<clasnfo>`**
 * [value-based sizeof](https://github.com/kaitai-io/kaitai_struct/issues/84#issuecomment-486649530): **`reclasnfo._sizeof`**

Choose the one that suits you best. In both cases, you can delete this manually created attribute `total_size`, the size of the whole `clasnfo` type will be calculated automatically.

File: image/nitf.ksy [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+      has_mask:
+        value: image_sub_header.img_compression.substring(0, 1) == 'M' or image_sub_header.img_compression.substring(1, 2) == 'M'
+    seq:
+      - id: image_sub_header
+        type: image_sub_header
+      - id: image_data_mask
+        type: image_data_mask
+        if: has_mask
+      - id: image_data_field
+        size: _parent.header.linfo[idx].li.to_i - image_data_mask.total_size
```
Comment (generalmimon) [inline image/nitf.ksy:267]: Parsing of the file `rgb16.ntf` you provided in https://github.com/kaitai-io/kaitai_struct_formats/pull/294#issuecomment-653637125 fails due to `image_data_mask` being `null` (or more specifically, `undefined` in JavaScript - tested on https://ide.kaitai.io/devel/), but you're unconditionally trying to read `total_size` attribute of it. Shouldn't it have `if: has_mask` as well?

**Comment by rmspeers** [comment]
@generalmimon  / @KOLANICH - let me know if you see any other issues to resolve still.

File: image/nitf.ksy [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+        type: clasnfo
+      - id: resshl
+        type: str
+        size: 4
+      - id: resshf
+        type: str
+        size: resshl.to_i
+      - id: resdata
+        type: str
+        size: total_size - (2 + 25 + 2 + reclasnfo._sizeof + 4 + resshl.to_i)
```
Comment (generalmimon) [inline image/nitf.ksy:755]: @rmspeers Is there any reason not to use [`size-eos: true`](https://doc.kaitai.io/user_guide.html#_variable_length_structures) instead? It is intended specifically for this use case. I can add some quick suggestions for you to see what I mean.

File: image/nitf.ksy [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+        type: u2
+    seq:
+      - id: reserved_sub_header
+        type: reserved_sub_header(_parent.header.lrnfo[idx].lrsh.to_i)
+      - id: reserved_data_field
+        size: _parent.header.lrnfo[idx].lr.to_i
+  reserved_sub_header:
+    params:
+      - id: total_size
+        type: u2
```
Comment (generalmimon) [inline image/nitf.ksy:735]:
```suggestion
```

File: image/nitf.ksy [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+      - id: sfh_l2
+        type: str
+        size: 7
+  reserved_extension_segment:
+    params:
+      - id: idx
+        type: u2
+    seq:
+      - id: reserved_sub_header
+        type: reserved_sub_header(_parent.header.lrnfo[idx].lrsh.to_i)
```
Comment (generalmimon) [inline image/nitf.ksy:729]:
```suggestion
        type: reserved_sub_header
        size: _parent.header.lrnfo[idx].lrsh.to_i
```

File: image/nitf.ksy [hunk 11]
```diff
@@ -0,0 +1,766 @@
...
+        size: 4
+      - id: desshf
+        type: str
+        size: des_defined_subheader_fields_len.to_i
+      - id: des_defined_data_field
+        -orig-id: desdata
+        type: str
+        size: >
+          total_size - (2 + 25 + 2 + des_base.declasnfo._sizeof + 
+            (tre_ofl ? 6 + 3 : 0) + 4 + des_defined_subheader_fields_len.to_i)
```
Comment (generalmimon) [inline image/nitf.ksy:678]: Again, the same thing as in https://github.com/kaitai-io/kaitai_struct_formats/pull/294#discussion_r449780897.

File: image/nitf.ksy [hunk 12]
```diff
@@ -0,0 +1,766 @@
...
+        size: 2
+      - id: declasnfo
+        type: clasnfo
+  data_sub_header:
+    instances:
+      tre_ofl:
+        value: des_base.desid == 'TRE_OVERFLOW'
+    params:
+      - id: total_size
+        type: u2
```
Comment (generalmimon) [inline image/nitf.ksy:652]:
```suggestion
```

File: image/nitf.ksy [hunk 13]
```diff
@@ -0,0 +1,766 @@
...
+        size: 3
+      - id: text_extended_sub_header
+        type: tre_header
+  data_extension_segment:
+    params:
+      - id: idx
+        type: u2
+    seq:
+      - id: data_sub_header
+        type: data_sub_header(_parent.header.ldnfo[idx].ldsh.to_i)
```
Comment (generalmimon) [inline image/nitf.ksy:631]:
```suggestion
        type: data_sub_header
        size: _parent.header.ldnfo[idx].ldsh.to_i
```

See https://github.com/kaitai-io/kaitai_struct_formats/pull/294#discussion_r449781522

File: image/nitf.ksy [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+    - ntf.r2
+    - ntf.r3
+    - ntf.r4
+    - ntf.r5
+  ks-version: 0.9
+  encoding: UTF-8
+  endian: be
+  license: MIT
+doc: |
+  Is set to version format (file_version) of 02.10 and standard_type of BF01. Implemented by River Loop Security.
```
Comment (generalmimon) [inline image/nitf.ksy:18]: @rmspeers Can you please add a few sentences describing the format here? What it is used for? What kind of data it stores? What is the main usage field?

The point is for someone who visits the format page in our format gallery (e.g. [ASN.1 DER](https://formats.kaitai.io/asn1_der/index.html) format) to get a basic idea about what kind of the format it is, and not have to search it on Wikipedia or other external websites.

Comment (rmspeers) [reply image/nitf.ksy]: Added, see if that is what you were looking for?

File: image/nitf.ksy [hunk 15]
```diff
@@ -0,0 +1,769 @@
...
+      - id: srel
+        type: str
+        size: 20
+      - id: sdctp
+        type: str
+        size: 2
+      - id: sdcdt
+        type: str
+        size: 8
+      - id: sdcxm
```
Comment (generalmimon) [inline image/nitf.ksy:192]: I see that you improved a lot of field namings in https://github.com/kaitai-io/kaitai_struct_formats/pull/294/commits/77c94a25c82c78c2335fb6d888102dde70a649e3, but I still feel that these (and similar ones) deserve a more meaningful `id`.

File: image/nitf.ksy [hunk 16]
```diff
@@ -0,0 +1,769 @@
...
+  endian: be
+  license: MIT
+doc: |
+  The NITF (National Image Transition Format) format is a file format developed by the U.S. Government for
+  storing imagery, e.g. from satellites.
+
+  According to the [foreword of the specification](https://gwg.nga.mil/ntb/baseline/docs/2500c/2500C.pdf):
+  > The National Imagery Transmission Format Standard (NITFS) is the suite of standards for formatting digital
+  > imagery and imagery-related products and exchanging them among members of the Intelligence Community (IC) as
+  > defined by the Executive Order 12333, and other United States Government departments and agencies."
```
Comment (generalmimon) [inline image/nitf.ksy:24]: @rmspeers A **lot** better! 👍

File: image/nitf.ksy (hunk 14 above)
Comment (generalmimon) [reply image/nitf.ksy]: > Added, see if that is what you were looking for?

Yes, that's better 👍

File: image/nitf.ksy (hunk 15 above)
Comment (rmspeers) [reply image/nitf.ksy]: Done

Comment (rmspeers) [reply image/nitf.ksy]: Anything else we need to do or is this good to merge?

File: image/nitf.ksy [hunk 17]
```diff
@@ -0,0 +1,819 @@
...
+  file-extension:
+    - ntf
+    - nitf
+    - ntf.r0
+    - ntf.r1
+    - ntf.r2
+    - ntf.r3
+    - ntf.r4
+    - ntf.r5
+  ks-version: 0.9
```
Comment (generalmimon) [inline image/nitf.ksy:13]: Now that we got rid of all `_sizeof` usages in commit https://github.com/kaitai-io/kaitai_struct_formats/pull/294/commits/fb0333d5d21ca782295f00b571373bce240a9e03, which was the only feature available from the (not yet released) 0.9 version, you can calmly lower the minimum KS compiler version needed to compile your spec to **0.8**.

```suggestion
  ks-version: 0.8
```

**Pull request merged**